### ISSUE TITLE
Allocations static flags issues.

### ISSUE THREAD

**Issue by njjc** [SalesforceFoundation/Cumulus#2591]
Allocations triggers have multiple issues due to improper static flags implementation: 
- they don't run correctly after workflow field updates on opportunity
- they don't run correctly in when opportunity update DML happens before `after insert` triggers have a chance to fire

**Comment by njjc** [comment]
**lurch: add

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x80000006VCWFAA4" target="_blank">W-018818</a>

**Comment by judisohn** [comment]
The most widely reported symptom of this is when GAU Allocations are not added to Opportunities that have a Primary Campaign Source with GAU Allocations and there is another app or code that is running on Opportunity update. We have also seen this where a workflow updates a field on a record, such as Recurring Donations's Next Payment Date, and the trigger that should be firing to update opportunities does not fire.

**Comment by force2b** [comment]
@judisohn 
I was able to reproduce this issue with a simple WFR on the Opportunity:

- Set up a default GAU for new Opportunities
- Create a WFR on Opportunity for "Create" that populates the Amount field when the field is blank.
- Create an Opp with a blank amount field
- Expected: The Amount field has a value and there is a default Allocation created
- Actual: The Amount has a value, but there is no default Allocation created.

I wasn't able to reproduce the issue with an After Insert or After Update trigger, nor with Primary Campaign Source. Do you have any additional steps on this that I can follow?

**Comment by njjc** [comment]
For the Primary Campaign Source, have a campaign with a related campaign allocation, then use a workflow rule to set the source after insert.

For the update before insert, you'll need a trigger outside of the TDTM framework that happens to run before the TDTM trigger that updates the current opportunity trigger set (cloned or re-created with the same ID).

**Comment by force2b** [comment]
Thanks. The key was using the default Allocation on the Primary Campaign. Once I set that up (rather than use the NPSP Setting for "Create a Default Allocation"), I was able to reproduce the issue where the default allocation is not created for the Opportunity.

The test After Insert Opportunity trigger will execute before the TDTM "After Insert" trigger. In that trigger, if the Amount is null, it creates a new Opportunity Sobject with the Amount field to 100 and commits that to the db. Debug logs show that the execution path as follows:

1. NPSP Before Insert (Opportunity) Trigger
1. Custom After Insert (Opportunity) Trigger (DML occurs to set Amount to 100)
1. NPSP Before Update (Opportunity) Trigger
1. NPSP After Update (Opportunity) Trigger
1. NPSP Before Update (Account) Trigger
1. NPSP After Update (Account) Trigger
1. NPSP After Insert (Opportunity) Trigger

When Primary Campaign is not the determination for the default allocation, the "After Update Opportunity" trigger (step 4 above) is what will create the default allocation when it sees that the Amount was changed from null to 100. That doesn't seem to work when a Primary Campaign Source is involved.
